src/transformStateWithClones.js

```diff
@@ -7,7 +7,34 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateVersions = [];
+  let currentState = state;
+
+  for (const action of actions) {
+    let nextState;
+
+    if (action.type === 'clear') {
+      nextState = {};
+    } else {
+      nextState = { ...currentState };
+
+      if (action.type === 'removeProperties') {
+        for (const key of action.keysToRemove) {
+          delete nextState[key];
+        }
+      }
+
+      if (action.type === 'addProperties') {
+        nextState = {
+          ...nextState, ...action.extraData,
+        };
+      }
+    }
+    stateVersions.push(nextState);
+    currentState = nextState;
+  }
+
+  return stateVersions;
 }
```

Comment:
Good work but let's rewrite it using switch/case since we rely on a single action.type value